### PR TITLE
Simplify API Usage & Data Flow

### PR DESCRIPTION
**Pull request by erikquinn**
Should have discussed all the concepts already more or less. Peruse at will.

Note that this **is not complete**-- you will probably be best equipped to finish it. In `renderPolygons.tsx`, the new data source is available as `eqGroupData` on line 124, but I dared not mess with the giant nested mess that came after it. At least not at this hour. Once that is connected, the other remaining deprecated api routes and some other logic could then be removed.

I can have another go at it in a few days if you don't have the time. Let me know if any questions/concerns.

### CODE REVIEW

File: public/facilityData/zmaSectors.json [hunk 1]
```diff
@@ -31,11 +31,6 @@
               [-80.98611111, 28.72777778],
               [-80.98333333, 28.66666667],
               [-80.430556, 27.306111],
-              [-80.3825, 27.133611],
-              [-80.22166667, 27.15],
-              [-80.06666667, 27.15],
-              [-79.95, 27.17083333],
-              [-79.83910896, 27.25185306],
```
Comment (erikquinn) [inline public/facilityData/zmaSectors.json:38]: whoops, I guess this can be removed

File: prisma/schema.prisma [hunk 2]
```diff
@@ -91,6 +91,18 @@ model FIR {
...
 }
 
+// Client Deliverables
+model Deliverables {
+  deliverableName String  @unique // "groupData" or "datasetTimestamps"
+  content         String          // stringified json to send to the client
+}
+
+// Deliverable Timestamps
+model Timestamps {
```
Comment (David-Fryd) [inline prisma/schema.prisma:101]: Should be hashes/keys not timestamps, to represent versioning. Client should store the hash locally in state, and use *that* for version comparison. Timestamps will constantly change and for our purposes a single version hash/id is sufficient.

File: package.json [hunk 3]
```diff
@@ -24,6 +24,7 @@
     "@turf/turf": "^6.5.0",
     "@types/leaflet": "^1.9.3",
     "leaflet": "^1.9.3",
+    "lodash": "^4.17.21",
```
Comment (David-Fryd) [inline package.json:27]: What exactly are we using lodash for?

File: src/components/TempSplitter.tsx [hunk 4]
```diff
@@ -151,6 +151,10 @@ export const TempSplitter = ({ firName }: { firName: string }) => {
       firName: firName,
     });
 
+  // TODO: We have already retrieved the latest facility data in MainMap.tsx
+  //       Do not make any new requests here, just figure out how to share
+  //       the data from the MainMap to the TempSplitter. Is there somewhere
```
Comment (David-Fryd) [inline src/components/TempSplitter.tsx:156]: TempSplitter has been entirely removed on the splitconfig branch in place of the actual page, this logic was just temporary for sake of the ZDC demo. 

As to the larger point of using context or something else to pass down data, I don't actually think we need to use context. The MainMap should reflect the latest data, and continue to refresh and fetch if it has outdated information. As for the splitconfig page, we *will* need to make another request on every fresh render of the splitconfig page because we want to show what the latest group data is whenever I go in and begin to edit a split. I won't however want that page to constantly refresh as I am regrouping sectors, which would be quite frustrating as a user. This means meaning that we actually don't want to reference a master context with constantly-updating data. If this is the case, there is no other component besides MainMap where this data would be relevant, and in the interest of not over-utilizing context, we can probably isolate the fetch and refresh logic to that component. If there are other components that require this constantly up-to-date data, then context might be appropriate.

This data would only change, by the way, when controllers are simultaneously attempting to edit the groupings of the same facility. The consequences of this are that changes made by the latest-submitting user will override the grouping of the user who submitted first, but I think that is acceptable and expected behavior if you are editing the groupings of the same facility.

File: src/utils/facilityData/server/syncFacilityData.ts [hunk 5]
```diff
@@ -30,7 +32,8 @@ export async function syncFacilityData({
   const directoryPath = path.join(process.cwd(), "public", "facilityData");
 
   // List all files in the directory
-  const files = await fs.readdir(directoryPath);
+  let files = await fs.readdir(directoryPath);
+  files = _.without(files, '~allFacilities.json');
```
Comment (David-Fryd) [inline src/utils/facilityData/server/syncFacilityData.ts:36]: We should no longer be placing the individual facilityData files inside of the `/public/` directory because we should not be serving the individual files at all. Instead, we should relocate the files to a `/facilityData/` directory in the root of the project, and only the generated `allFacilities.json` file should appear in `/public/allFacilities` (we could probably remove the `/facilityData` directory from public since `allFacilityData` would be be the only file there anyway).

In the same vein, perhaps removing the unnecessary `~` prefix to the filename, which might suggest its a temporary file when in reality its a generated file intended for long-term use.

File: src/types/facilityData.ts [hunk 6]
```diff
@@ -25,12 +25,63 @@ export type Geojson = {
...
+  notes: string,
+  color: string,
+  sectorList: string[]
+}
+
+// TODO: Once everything is working, just use this format, and reformat the data files to match it.
+// facility shape used in comprehensive facility file
+export interface FacilityImproved {
+  firDetails: {
+    id: number, // TODO: Do we need this id field? Presumably not? Why did I put it there?
```
Comment (David-Fryd) [inline src/types/facilityData.ts:65]: Not sure! Any idea?

File: src/types/facilityData.ts [hunk 7]
```diff
@@ -25,12 +25,63 @@ export type Geojson = {
...
+    lid: string,
```
Comment (David-Fryd) [inline src/types/facilityData.ts:66]: Not sure what `lid` represents?

File: src/utils/facilityData/server/syncFacilityData.ts
Comment (David-Fryd) [inline src/utils/facilityData/server/syncFacilityData.ts:1]: We should ensure that we also somehow invoke this logic (or similar) in the `npm run build` pipeline, such that data is synced at build time.

File: src/server/api/routers/facilitydata.ts [hunk 8]
```diff
@@ -33,22 +21,33 @@ export const facilityDataRouter = createTRPCRouter({
...
-  getFacilityDataFilenames: publicProcedure.query(({ ctx }) => {
-    const directoryPath = path.join(process.cwd(), "public", "facilityData");
-    // List all files in the directory
-    const files = fs.readdir(directoryPath);
-    return files;
+  getGroupData: publicProcedure.query(({ ctx }) => {
+    return ctx.prisma.deliverables.findFirst({
+      where: { deliverableName: 'groupData' },
+    });
   }),
```
Comment (David-Fryd) [inline src/server/api/routers/facilitydata.ts:39]: I acknowledge the amount of TODOs here, this will all change in the splitconfig branch soon to support the drag-n-drop functionality.

**Review by David-Fryd — CHANGES_REQUESTED** [review]

File: src/components/TempSplitter.tsx (hunk 4 above)
Comment (erikquinn) [reply src/components/TempSplitter.tsx:156]: > As for the splitconfig page, we will need to make another request on every fresh render of the splitconfig page because we want to show what the latest group data is whenever I go in and begin to edit a split.

Yes, agreed. Though I'm proposing an API request to be made to compare the version of the data we have vs the latest available about every 10 seconds at the longest. So it's unlikely it would be out of date in the first place. Regardless though, it should do a fresh query when the split page is opened just to be safe.

> I won't however want that page to constantly refresh as I am regrouping sectors

This i definitely do not agree with though. Everything should always display the latest state possible, for everything. I see no reason this should be an exception.

To give a more tangible context for that claim, borrowing from your example... If two people are editing ZMA's split configuration, in your plan, neither will know the other is doing anything, and whoever submits it last will have their changes applied. The first person could reconfigure, check the map, confirm it's correct and updated. The second person would do the same thing. When the first person's changes are overwritten, they won't know unless they're still viewing the map. If, instead, things are moving in the configuration page, it at least prompts the two conflicting users to go "hey, who's doing that?" and resolve to pick who will edit the configuration, because you never want two people editing it at once.

File: package.json (hunk 3 above)
Comment (erikquinn) [reply package.json:27]: I'm a lodash addict. I used a few in a few places... `_without()` (to find and remove a provided element from an array), and `_isEqual()` (a deep array comparison).

File: prisma/schema.prisma (hunk 2 above)
Comment (erikquinn) [reply prisma/schema.prisma:101]: Call it whatever you want 🤷‍♂️ 

I do think the date context is still potentially helpful, and it works as a guaranteed-unique value. For clarity, are you saying to rename it, or to not use a Date.now() value?

File: src/types/facilityData.ts (hunk 7 above)
Comment (erikquinn) [reply src/types/facilityData.ts:66]: Thought I had a comment there... It's an established and specific term, actually. Though not particularly widely known. Is an FAA term, Location IDentifier. Each FAA facility has one. Washington ARTCC's lid is `ZDC`.

Currently you are calling ZDC the `.name`, which is rather misleading.

File: src/utils/facilityData/server/syncFacilityData.ts
Comment (erikquinn) [reply src/utils/facilityData/server/syncFacilityData.ts:1]: In fairness, it would be easy enough to just go press the button when deploying a new version... But if we can hook it up to the build script, then by all means!

File: src/server/api/routers/facilitydata.ts (hunk 8 above)
Comment (erikquinn) [reply src/server/api/routers/facilitydata.ts:39]: Sounds good. I leave todo litter most places I go 😜

File: prisma/schema.prisma (hunk 2 above)
Comment (David-Fryd) [reply prisma/schema.prisma:101]: Ah, I think I misunderstood the way in which the version-comparison would work, which perhaps suggests we could find a more descriptive/clear name (or document it more thoroughly).

I initially thought that two different times were being compared to check version compatibility (time of check with the time of update, which didn't make much sense). What we intend to do is pass that timestamp to the client and then have the client use that as the "last retrieved info updated at _X_ from server, does server have anything > _X_?"

Is that understanding correct?

File: src/components/TempSplitter.tsx (hunk 4 above)
Comment (David-Fryd) [reply src/components/TempSplitter.tsx:156]: The solution to the problem you proposed is not to provide constant real-time facilityData and groupData updates, but to have an external system (a different system) for tracking and detecting whether another user is on the page making edits or not. 

Assuming we did provide the constant updates for facility/group data, there are two big issues:
1) The UI shifting around in the middle of a configuration session is terrible for UX because it's confusing, frustrating, and would lead to mistakes being made. Consider the case where a group you are in the middle of editing suddenly disappears because another user deleted it. Consider another case where you don't notice that sector *X* was automatically and quietly moved from group "B" to "C" because you were focusing on group "E" at the time. Maybe you go back and are confused when you notice the inconsistency, maybe you don't catch it. Either way is a very bad thing for UX's sake. The only time where this type of behavior is reasonable is when it comes to real-time collaboration and you can explicitly see who is making changes and to what, which we don't support.

2) Even if we provided the same type of data that the MainMap receives to the splitconfig page, it still doesn't prevent the case where two users don't realize they are editing the same thing until its too late. Only when one person is done and submits would the other user realize that edits have been made, at which point both users are not only deep into configuration, but also means that the first user might not check back after submitting, and have their changes overwritten anyway. 

In summary, I don't think we need to expose the constantly changing group/facility data, though we might need a different system for handling editing conflicts.

File: src/types/facilityData.ts (hunk 7 above)
Comment (David-Fryd) [reply src/types/facilityData.ts:66]: Ah. As long as its well documented its fine, however I might even encourage a more verbose name for clarity's sake? Especially since capitalization is inconsistent among the schemas/type files/trpc representation, etc...

File: prisma/schema.prisma (hunk 2 above)
Comment (erikquinn) [reply prisma/schema.prisma:101]: Essentially, though in reverse. My thoughts was:
- client makes a request for new data, and includes as a part of that request a simple object like `{ facilityData: 12345, group data: 34567 }`, to which the API would respond with:
    - "up to date", or
    - new group data set, or
    - new facility data set, or
    - new group + facility data sets

Comment (erikquinn) [reply prisma/schema.prisma:101]: ...noting that the data sets include the timestamp within them. So by the API sending new group data, the client will then make their next request using that timestamp.

File: src/components/TempSplitter.tsx (hunk 4 above)
Comment (erikquinn) [reply src/components/TempSplitter.tsx:156]: I don't really know what to say other than... I totally disagree.

Having "an external system for tracking and detecting whether another user is on the page making edits" may be a handy addition that could be added post MVP. Even then, I would say this would objectively be of very low value. But for the time being, we don't need any of it.

You're sort of blowing the "problem" out of proportion and prioritizing the wrong things. Cases where multiple people are simultaneously editing a split will always be extremely rare. And you are pushing for a system that gives priority for handling this rare scenario, at the expense of behavior in the 99.9% of cases where only one person is editing. Every single thing you mention in point 1 is somewhere between acceptable behavior and actually_desirable behavior_. Not one of those behaviors is even remotely a negative. I understand you have experience in software and UX that says "this is not how this should work", and you perceive the problem as being significant. But I need to have some degree of latitude to say what is and is not important for the people who would use the tool, as I am the one with the experience of what is important to the center controllers, TMUs, Command Center, etc, and how the tool should behave for them. To that end, I am saying that the protections you're trying to establish for simultaneous edits _are not remotely important_-- the things you list as problems are not problems at all, whereas protecting the simple concept of "anything you ever see anywhere is always completely up to date at all times for everyone" is critically important. We should not degrade the behavior under normal conditions for an edge-case scenario which has no downsides.

File: public/facilityData/zmaSectors.json (hunk 1 above)
Comment (David-Fryd) [reply public/facilityData/zmaSectors.json:38]: the commit should be removed? or the data removed within the file should stay removed?
